infrastructure/authentication/auth_service_impl.go

```diff
@@ -135,32 +144,99 @@ func (a *AuthenticationServiceImpl) IsAuthenticated() bool {
 
 		invalidToken, isLegacyTokenErr := a.c.TokenAsOAuthToken()
 
-		// we always log out
-		logger.Debug().Msg("logging out")
-		a.Logout(context.Background())
-
-		// determine the right error message
-		if isLegacyTokenErr == nil {
-			// it is an oauth token
-			if invalidToken.Expiry.Before(time.Now()) {
-				a.handleFailedRefresh()
-			} else {
-				// access token not expired, but creds still not work
-				a.HandleInvalidCredentials()
-			}
+		if logoutCausingError(err) {
+			a.handleLogoutCausingError(logger, isLegacyTokenErr, invalidToken)
+			return false
 		} else {
-			// legacy token does not work
-			a.HandleInvalidCredentials()
+			// try again
+			time.Sleep(2 * time.Second)
+			retryUser, retryError := a.provider.GetCheckAuthenticationFunction()()
+			if retryUser == "" || retryError != nil {
+				// retry failed again, we gotta logout after all
+				a.handleLogoutCausingError(logger, isLegacyTokenErr, invalidToken)
+				return false
+			}
 		}
-		return false
 	}
-
 	// we cache the API auth ok for up to 1 minutes after last access. Afterwards, a new check is performed.
 	a.authCache.Set(a.c.Token(), true, imcache.WithSlidingExpiration(time.Minute))
 	a.c.Logger().Debug().Msg("IsAuthenticated: " + user + ", adding to cache.")
 	return true
 }
 
+func (a *AuthenticationServiceImpl) handleLogoutCausingError(logger zerolog.Logger, isLegacyTokenErr error, invalidToken oauth2.Token) {
+	logger.Debug().Msg("logging out")
+	a.Logout(context.Background())
+
+	// determine the right error message
+	if isLegacyTokenErr == nil {
+		// it is an oauth token
+		if invalidToken.Expiry.Before(time.Now()) {
+			a.handleFailedRefresh()
+		} else {
+			// access token not expired, but creds still not work
+			a.HandleInvalidCredentials()
+		}
+	} else {
+		// legacy token does not work
+		a.HandleInvalidCredentials()
+	}
+}
+
+func logoutCausingError(err error) bool {
+	if err == nil {
+		return true
+	}
+
+	// Check for context cancellation
+	if errors.Is(err, context.Canceled) {
+		return false
+	}
+
+	// Check for timeout errors
+	if errors.Is(err, context.DeadlineExceeded) {
+		return false
+	}
+
+	// Check for URL errors
+	var urlErr *url.Error
+	if errors.As(err, &urlErr) {
+		return false
+	}
+
+	// Check for network operation errors
+	var netErr *net.OpError
+	if errors.As(err, &netErr) {
+		return false
+	}
+
+	// Check for DNS errors
+	var dnsErr *net.DNSError
+	if errors.As(err, &dnsErr) {
+		return false
+	}
+
+	// Check for system call errors
+	var sysCallErr *os.SyscallError
+	if errors.As(err, &sysCallErr) {
+		return false
+	}
+
+	// Check for connection reset error
+	if errors.Is(err, syscall.ECONNRESET) {
+		return false
+	}
+
+	// Check for EOF error
+	if errors.Is(err, io.EOF) {
+		return false
+	}
+
+	// as we can't enforce correct error reporting, let's do a final check on the internet connection, whether it's there
+	_, err = http.Get("https://www.google.com")
```

Comment:
I'm not sure if this will be ok. It will depend on customer's infra setup. 
I don't suppose they would block google.com but it could happen.

Maybe we could instead try to HTTP GET config.SnykUI() ?

Reply:
Good idea!